### PR TITLE
PAM authentication implementation

### PR DESCRIPTION
**Pull request by lakshmi-kannan**

### CODE REVIEW

**Comment by lakshmi-kannan** [comment]
Merging! Comment if you need something fixed.

File: tests/unit/test_authenticate.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
```
Comment (Kami) [inline tests/unit/test_authenticate.py]: unit -> integration :P

**Comment by Kami** [comment]
I wanted to say it would also be good to add a test case which tests multiple failed login attemps and tally counter, but it looks like that change got reverted in salt (https://github.com/saltstack/salt/commit/b957ea8977b9c469d32dab64ea4dafac02dc7d1c, https://github.com/saltstack/salt/pull/22895#ref-commit-b957ea8) and I can't find any reference why, sigh :/

**Comment by Kami** [comment]
Another thing I think it would be good to do to reduce the potential attack surface area is for StackStorm pam auth to only works for users which are part of a special group we created.

I know this is more related to the deployment (would probably also need to modify / write a custom pam config) than this package, but still something to thing about.

**Comment by manasdk** [comment]
Shouldn't we handled multiple failed logins at st2 level? This way each adapter does not have to deal with this bit. Is there something that makes it adapter specific?

**Comment by manasdk** [comment]
> Another thing I think it would be good to do to reduce the potential attack surface area is for StackStorm pam auth to only works for users which are part of a special group we created.

Can you explain the surface? I am not able to work what this guards against.

**Comment by Kami** [comment]
@manasdk 

> Shouldn't we handled multiple failed logins at st2 level? This way each adapter does not have to deal with this bit. Is there something that makes it adapter specific?

Ideally yeah, but this means more code and features we need to implement and maintain so we should try to outsource it.

> Another thing I think it would be good to do to reduce the potential attack surface area is for StackStorm pam auth to only works for users which are part of a special group we created.

IIRC, right now any system user can log into StackStorm with pam auth (e.g. if a user adds a system / linux account that user will be able to log into StackStorm, not to mention that there are already a bunch of different service, etc. account present on the system. Luckily most of them don't allow login so those shouldn't be able to login into StackStorm).

**Comment by manasdk** [comment]
> Ideally yeah, but this means more code and features we need to implement and maintain so we should try to outsource it.

Alright. In theory I agree with writing less code but not in this case. Blocking on multiple failed logins needs to be consistently applied in StackStorm across all adapters so I would say it is code we need to up level

**Comment by manasdk** [comment]
> IIRC, right now any system user can log into StackStorm with pam auth

Isn't that the point if PAM? I am OK with reducing the scope, seems like a good idea to be restrictive.

**Comment by lakshmi-kannan** [comment]
I am fine with scoping the access to only users in a certain group.

File: tests/unit/test_authenticate.py (hunk 1 above)
Comment (lakshmi-kannan) [inline tests/unit/test_authenticate.py]: https://github.com/StackStorm/st2-auth-backend-pam/pull/3

**Pull request merged**